cpp/arcticdb/column_store/key_segment.cpp

```diff
@@ -0,0 +1,146 @@
+/* Copyright 2024 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#include <arcticdb/column_store/key_segment.hpp>
+#include <arcticdb/pipeline/index_fields.hpp>
+
+namespace arcticdb {
+
+KeySegment::KeySegment(SegmentInMemory&& segment, SymbolStructure symbol_structure):
+        num_keys_(segment.row_count()),
+        string_pool_(segment.string_pool_ptr()),
+        symbol_structure_(symbol_structure)
+{
+    // Needed as the column map is not initialised at read time if the segment has no rows
+    segment.init_column_map();
+    stream_ids_ = segment.column_ptr(static_cast<uint32_t>(pipelines::index::Fields::stream_id));
+    version_ids_ = segment.column_ptr(static_cast<uint32_t>(pipelines::index::Fields::version_id));
+    creation_timestamps_ = segment.column_ptr(static_cast<uint32_t>(pipelines::index::Fields::creation_ts));
+    content_hashes_ = segment.column_ptr(static_cast<uint32_t>(pipelines::index::Fields::content_hash));
+    start_indexes_ = segment.column_ptr(static_cast<uint32_t>(pipelines::index::Fields::start_index));
+    end_indexes_ = segment.column_ptr(static_cast<uint32_t>(pipelines::index::Fields::end_index));
+    key_types_ = segment.column_ptr(static_cast<uint32_t>(pipelines::index::Fields::key_type));
+
+    switch (symbol_structure_) {
+        case SymbolStructure::SAME:
+            debug::check<ErrorCode::E_ASSERTION_FAILURE>(check_symbols_all_same(),
+                                                         "Expected all symbols to be identical in KeySegment");
+            if (stream_ids_->row_count() != 0) {
+                if (is_sequence_type(stream_ids_->type().data_type())) {
+                    symbol_ = std::string(string_pool_->get_const_view(stream_ids_->template reference_at<uint64_t>(0)));
+                } else {
+                    symbol_ = safe_convert_to_numeric_id(stream_ids_->template reference_at<uint64_t>(0));
+                }
+            }
+            break;
+        case SymbolStructure::UNIQUE:
+            debug::check<ErrorCode::E_ASSERTION_FAILURE>(check_symbols_all_unique(),
+                                                         "Expected all symbols to be unique in KeySegment");
+        case SymbolStructure::UNKNOWN:
+        default:
+            break;
```

Comment:
What happens when we don't recognize the symbol structure? Is this expected?

Reply:
That's fine, it just means optimisations available to the other 2 enum values can't be applied (not that they're generally implemented yet).
This might get replaced with inheritance at some point, but this is sufficient for now.